### ISSUE TITLE
Exchange bindings stop working after broker restart

### ISSUE THREAD

**Issue by cristian-moreno-ruiz** [rabbitmq/rabbitmq-delayed-message-exchange#149]
### Description

I am working on an application where we use RabbiMQ with rabbitmq_delayed_message_exchange plugin installed, and we use an 'x-delayed-message' exchange (with 'x-delayed-type: topic') where we send all messages, which are later routed to one of our queues, based on the routingKey. The setup works fine when we create it, but after we restart the broker, new messages are not routed any more.

### Platforms
I've been able to reproduce this issue on those setups:
1. Running Docker on mac, from rabbitmq:3-management:
RabbitMQ 3.8.5
Erlang 23.0.2
Plugin (rabbitmq_delayed_message_exchange) 3.8.0
2. Running the server locally in my macOS Catalina:
RabbitMQ 3.8.6
Erlang 23.0.3
Plugin (rabbitmq_delayed_message_exchange) 3.8.0

### Client Libraries
I am able to reproduce the issue both using the management UI/CLI (with the management plugin enabled), and with the NodeJS amqplib package.

#### Through the management UI
Using RabbitMQ server, with rabbitmq_delayed_message_exchange:
1. Add a new exchange with
    - type: x-delayed-message
    - durability: durable
    - arguments: x-delayed-type: topic
2. Create a Classic Queue, with Durability: Durable.
3. Create a binding in the exchange created at (1) to route messages to the queue created at (2), specifying a routingKey.
4. Publish some message to the exchange, providing the same routingKey used in (3)
    - Message should be routed successfully to the queue.
5. Restart the server
6. Repeat step (4)
    - Messages are not routed any more

#### Through the management CLI, commands to reproduce (above steps)
$ rabbitmqadmin declare exchange name=exchange durable=true type=x-delayed-message 'arguments={"x-delayed-type":"topic"}'
$ rabbitmqadmin declare queue name=queue durable=true queue_type=classic

$ rabbitmqadmin declare binding source=exchange destination=queue routing_key=queue

$ rabbitmqadmin publish routing_key=queue exchange=exchange payload=test

> Message published

 (Now we restart the server)

$ rabbitmqadmin publish routing_key=queue exchange=exchange payload=test

> Message published but NOT routed

### How to Fix it
When this happens, you can delete the binding created in step (3), and then repeat step (3). Now, new messages created like in (4), will be routed properly again.

It is actually not possible to fix it from the CLI, because trying to delete results in  "Not found" (maybe related with the issue), and redeclaring the binding doesn't fix it neither:

$ rabbitmqadmin delete binding source=exchange destination_type=queue destination=queue

> *** Not found: /api/bindings/%2F/e/exchange/q/queue/~

### Additional Notes:

Some of my colleagues have an older version of RabbitMQ installed and can't reproduce the issue:
1. Running Docker on mac, from rabbitmq:3-management:
RabbitMQ 3.8.2
Erlang 22.2.2
Plugin (rabbitmq_delayed_message_exchange) 3.8.0
2. Running the server locally in macOS Catalina:
RabbitMQ 3.8.2
Erlang 23.0.3
Plugin (rabbitmq_delayed_message_exchange) 3.8.0

Finally, I could also not reproduce the mentioned issue when using an x-delay-message exchange with `x-delayed-type: direct`, this will only happen with `topic`.

I appreciate any help on this, an I would be glad to clarify any of the steps to reproduce the issue.

Link to initial Google Groups discussion: https://groups.google.com/g/rabbitmq-users/c/mbS7gokR6YA/m/HFMPdzjDCgAJ

**Comment by kjnilsson** [comment]
I have tested this and confirmed it - for some reason the routing that is performed after the delay timer triggers results in no matches for the binding.

**Comment by kjnilsson** [comment]
It looks like the rabbit_topic_trie_* mnesia tables are not re-populated for the delayed message exchange bindings after a restart.

**Comment by michaelklishin** [comment]
Sounds like https://github.com/rabbitmq/rabbitmq-consistent-hash-exchange/issues/45 in a different custom exchange type.

**Comment by kjnilsson** [comment]
Yes could well be.

**Comment by michaelklishin** [comment]
We can confirm that this is a yet another casualty of the fundamental chicken-and-egg problem described in https://github.com/rabbitmq/rabbitmq-server/issues/2384. What works for some cases and plugins doesn't necessarily help stateful exchange types such as `rabbitmq_consistent_hash_exchange` and this one.

**Comment by michaelklishin** [comment]
https://github.com/rabbitmq/rabbitmq-consistent-hash-exchange/pull/46 demonstrates what `rabbitmq_consistent_hash_exchange` had to do to workaround the chicken-and-egg problem: avoid using boot steps and instead recover its state manually.

We cannot import definitions before all plugins are enabled, cannot enable some plugins without having all the schema rows (some of which may come from definitions) and cannot locate some boot steps without activating plugins. Instead of changing the core a few more times to accommodate a very small list of affected plugins, the plugins should adapt to the new reality and recover its state.

**Comment by michaelklishin** [comment]
I could reproduce.

**Comment by michaelklishin** [comment]
@cristian-moreno-ruiz can you please give https://github.com/rabbitmq/rabbitmq-delayed-message-exchange/releases/tag/3.8.9 a shot? I only tested it on `3.8.9`, so we highly recommend upgrading. Key changes that affect this plugin started with RabbitMQ `3.8.4`.

**Comment by BSchwetzel** [comment]
So we stumbled across this problem as well today for the second time, the first time around we couldn't attribute this problem properly. After we made sure that changing our exchange type from x-delayed-message to topic circumvents this problem (which happens in our clustered Docker Swarm deployment if the brokers don't restart in the correct order) we then found this issue and were very happy to find a release which aims to solve this problem.

And indeed it does :) We updated the plugin in our installation and can't reproduce this problem anymore. So thanks a ton @cristian-moreno-ruiz for bringing attention to this problem and @michaelklishin for fixing it (at least for us)!
